File: pubnub-gson/pubnub-gson-docs/src/main/java/com/pubnub/docs/publishAndSubscribe/signal/SignalOthers.java
package com.pubnub.docs.publishAndSubscribe.signal;

import com.pubnub.api.PubNubException;
import com.pubnub.api.UserId;
import com.pubnub.api.java.PubNub;
import com.pubnub.api.java.v2.PNConfiguration;
import com.pubnub.api.java.v2.entities.Channel;
import com.pubnub.api.models.consumer.PNPublishResult;

public class SignalOthers {
    private void signalMessage() throws PubNubException {
        // https://www.pubnub.com/docs/sdks/java/api-reference/publish-and-subscribe#signal-a-message-to-a-channel

        // snippet.signalMessage
        PNConfiguration.Builder configBuilder = PNConfiguration.builder(new UserId("yourUserId"), "demo");
        configBuilder.publishKey("demo");

        PNConfiguration pnConfiguration = configBuilder.build();
        PubNub pubnub = PubNub.create(pnConfiguration);
        Channel channel = pubnub.channel("myChannel");

        channel.signal("Hello everyone!")
                .customMessageType("text-message")
                .async(result -> {
                    result.onSuccess((PNPublishResult publishResult) -> {
                        Long timetoken = publishResult.getTimetoken(); // signal message timetoken
                    }).onFailure( (PubNubException exception) -> {
                        exception.printStackTrace();
                    });
                });
        // snippet.end
    }
}

File: pubnub-kotlin/pubnub-kotlin-docs/src/main/kotlin/com/pubnub/docs/publishAndSubscribe/signal/SignalOthers.kt
package com.pubnub.docs.publishAndSubscribe.signal

import com.pubnub.api.PubNub
import com.pubnub.api.UserId
import com.pubnub.docs.SnippetBase

class SignalOthers : SnippetBase() {
    private fun signalBasic() {
        // https://www.pubnub.com/docs/sdks/kotlin/api-reference/publish-and-subscribe#basic-usage-2

        // snippet.signalBasic
        val configBuilder = com.pubnub.api.v2.PNConfiguration.builder(UserId("myUserId"), "demo").apply {
            publishKey = "demo"
        }
        val pubnub = PubNub.create(configBuilder.build())
        val channel = pubnub.channel("myChannel")

        channel.signal(
            message = "This is a signal!",
            customMessageType = "text-message"
        ).async { result ->
            result.onFailure { exception ->
                println("Error while publishing")
                exception.printStackTrace()
            }.onSuccess { signalResult ->
                println("Signal sent, timetoken: ${signalResult.timetoken}")
            }
        }
        // snippet.end
    }
}
